<a href="https://colab.research.google.com/github/matt-balda/pln-2023/blob/master/exercicios/Entropy_Word.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

- Importação da biblioteca math para a utilzação do log na base 2

In [85]:
from math import log2

In [118]:
def calculator_entropy(word):
  entropy = 0.0
  word_length = len(word)
  letter_counts = {}
  for letter in word:
    letter_counts[letter] = word.count(letter)

  for letter in letter_counts:
    rel_freq = letter_counts[letter] / word_length
    if rel_freq > 0:
      entropy += -rel_freq * log2(rel_freq)

  return entropy

result1 = calculator_entropy('ARARAQUARA')
result2 = calculator_entropy('UNIPAMPA')
result3 = calculator_entropy('ALEGRETE')

print(f'ARARAQUARA: {result1: .2f}')
print(f'UNIPAMPA: {result2: .2f}')
print(f'ALEGRETE: {result3: .2f}')

ARARAQUARA:  1.69
UNIPAMPA:  2.50
ALEGRETE:  2.41


- A que possui maior entropia é a palavra UNIPAMPA com 2.50 e é a mais complexa também porque "UNIPAMPA" tem uma combinação de letras mais variadas e apenas com uma repetição letra "P".